Sync/Synchronizers/Bookmarks/BookmarksRepairRequestor.swift

```diff
@@ -0,0 +1,535 @@
+
+    // We've found an error that forces us to abort this entire repair cycle.
+    case aborted = "repair.aborted"
+}
+
+struct RepairResponse {
+    let collection: String
+    let request: String
+    let flowID: String
+    let clientID: String
+    let ids: [String]
+
+    static func fromJSON(args: JSON) -> RepairResponse {
+        return RepairResponse(collection: args["collection"].stringValue, request: args["request"].stringValue,
+                              flowID: args["flowID"].stringValue, clientID: args["clientID"].stringValue,
+                              ids: args["ids"].arrayValue.map { $0.stringValue })
+    }
+}
+
+struct RepairRequest {
+    let collection: String
+    let request: String
+    let flowID: String
+    let requestor: String
+    let ids: [String]
+
+    func toSyncCommand() -> SyncCommand {
+        let jsonObj: [String: Any] = [
+            "command": "repairRequest",
+            "args": [
+                [
+                "collection": collection,
+                "request": request,
+                "flowID": flowID,
+                "requestor": requestor,
+                "ids": ids
+                ]
+            ]
+        ]
+        return SyncCommand(value: JSON(object: jsonObj).stringValue()!)
+    }
+}
+
+private class AbortRepairError: MaybeErrorType {
+    let description: String
+    init(_ description: String) {
+        self.description = description
+    }
+}
+
+private class UnknownClientError: MaybeErrorType {
+    let description: String
+    init(_ description: String) {
+        self.description = description
+    }
+}
+
+private class InvalidStateError: MaybeErrorType {
+    let description: String
+    init(_ description: String) {
+        self.description = description
+    }
+}
+
+class BookmarksRepairRequestor {
+    let prefs: Prefs
+    let remoteClients: RemoteClientsAndTabs
+    let scratchpad: Scratchpad
+
+    init(scratchpad: Scratchpad, basePrefs: Prefs, remoteClients: RemoteClientsAndTabs) {
+        self.scratchpad = scratchpad
+        self.prefs = basePrefs.branch("repairs.bookmark")
+        self.remoteClients = remoteClients
+    }
+
+    /**
+     * See if the repairer is willing and able to begin a repair process given
+     * the specified validation information.
+     *
+     * - returns: true if a repair was started and false otherwise.
+     */
+    func startRepairs(validationInfo: [(type: String, ids: [String])], flowID: String = Bytes.generateGUID()) -> Deferred<Maybe<Bool>> {
+        if (self.currentState != RepairState.notRepairing) {
+            log.info("Can't start a repair - repair with ID \(self.flowID) is already in progress")
+            return deferMaybe(false)
+        }
+
+        let ids = self.getProblemIDs(validationInfo)
+
+        if (ids.count == 0) {
+            log.info("Not starting a repair as there are no problems")
+            return deferMaybe(false)
+        }
+
+        if (ids.count > MaxRequestedIDs) {
```

Comment:
Nit: The `ifs` work well but it might be cleaner to use `guard` since you're checking for preconditions before executing the repair logic.